File: 백준/Silver/15649. N과 M （1）/N과 M （1）.java
import java.util.*;
import java.io.*;

public class Main {
    static int N, M;
    static int[] arr;
    static boolean[] visited;

    static void dfs(int cnt) {
        if(cnt == M) {
            for(int i = 0; i < M; i++) System.out.print(arr[i] + " ");
            System.out.println();
            return;
        }

        for(int i = 1; i <= N; i++) {
            if(!visited[i]) {
                visited[i] = true;
                arr[cnt] = i;
                dfs(cnt + 1);
                visited[i] = false;
            }
        }
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
        StringTokenizer st = new StringTokenizer(br.readLine());

        N = Integer.parseInt(st.nextToken());
        M = Integer.parseInt(st.nextToken());
        arr = new int[M];
        visited = new boolean[N + 1];

        dfs(0);
    }
}

File: 백준/Silver/15649. N과 M （1）/N과 M （1）.cc
#include <iostream>
#include <vector>
using namespace std;

int N, M;
int nums[9];
bool visited[9];

void dfs(int cnt) {
    if(cnt == M) {
        for(int i = 0; i < cnt; i++) cout << nums[i] << ' ';
        cout << '\n';
        return;
    }

    for(int i = 1; i <= N; i++) {
        if(!visited[i]) {
            visited[i] = true;
            nums[cnt] = i;
            dfs(cnt + 1);
            visited[i] = false;
        }
    }
}

int main(void) {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    cin >> N >> M;

    dfs(0);

    return 0;
}